Refactor for Driver interface

ChrisAubuchon opened the pull request:
- Move cloudapi calls into a `Driver` interface
- Create a mock driver for testing packer steps
- Create test cases for packer steps

ChrisAubuchon committed:
Add test cases

ChrisAubuchon committed:
Fix variable in test: label

ChrisAubuchon committed:
Change sdc_ variables to triton_

ChrisAubuchon committed:
Ignore vendor/ directory of packer repository

ChrisAubuchon committed:
Refactor to use a driver interface to enable test cases

ChrisAubuchon committed:
Remove uneeded file

ChrisAubuchon committed:
Add check for nil return in WaitFor commands.

If the cloudapi call fails nil is returned and the WaitFor commands
panic trying to dereference nil.

misterbisson commented on triton/access_config.go:
```diff
@@ -14,10 +14,10 @@ import (
+	Endpoint string `mapstructure:"triton_url"`
+	Account  string `mapstructure:"triton_account"`
+	KeyID    string `mapstructure:"triton_key_id"`
+	KeyPath  string `mapstructure:"triton_key_path"`
```
It appears to be `key_material` with the actual private key contents in the Terraform provider: https://github.com/hashicorp/terraform/blob/master/builtin/providers/triton/provider.go#L32. Which one is preferable?

ChrisAubuchon committed:
Change key_path to key_material

Determine if key_material is a file by calling `os.Stat()`. If it
is a valid file, read in the contents.

ChrisAubuchon committed:
More key_material changes

ChrisAubuchon committed:
Merge pull request #1 from asteris-llc/key_material

Key material

ChrisAubuchon committed:
cloudapi.Logger was removed. Create a new Logger for Triton client

ChrisAubuchon committed:
Remove the machine id from state when the machine is deleted

ChrisAubuchon committed:
Check for a machine id before deleting in cleanup

ChrisAubuchon committed:
Don't delete image as part of cleanup

misterbisson commented on example.json:
```diff
@@ -1,18 +1,18 @@
-        "sdc_account": "{{env `SDC_ACCOUNT`}}",
-        "sdc_key_id": "{{env `SDC_KEY_ID`}}",
-        "sdc_key_path": "{{env `SDC_KEY_PATH`}}"
+        "triton_url": "{{env `TRITON_URL`}}",
```
If this is intended to be compatible with the [Terraform provider](https://github.com/hashicorp/terraform/blob/master/builtin/providers/triton/provider.go#L18-L43) and other Joyent tools, it is probably preferable to use the `SDC_*` vars, even while calling them `triton_*` internally. The difference in naming is a frustration of history, but....

For context, [node-smartdc](https://github.com/joyent/node-smartdc#cli-setup-and-authentication) uses the following:
- `SDC_URL`
- `SDC_ACCOUNT`
- `SDC_KEY_ID`

[node-triton](https://www.joyent.com/blog/introducing-the-triton-command-line-tool), our new and improved CLI tool, will read those env vars as well as set them [from a profile](https://www.joyent.com/blog/introducing-the-triton-command-line-tool#using-profiles) with `triton env`, which is expected to be called as `eval "$(triton env)`. 

``` bash
$ triton env
export SDC_URL="https://us-sw-1.api.joyent.com"
export SDC_ACCOUNT="<my account>"
unset SDC_USER
export SDC_KEY_ID="SHA256:<my key>"
unset SDC_TESTING
export TRITON_PROFILE="us-sw-1"
# Run this command to configure your shell:
#     eval "$(triton env -st us-sw-1)"
```

I'm not sure if that means we should use `TRITON_KEY_MATERIAL` or `SDC_KEY_MATERIAL`, however, though the [Terraform provider](https://github.com/hashicorp/terraform/blob/master/builtin/providers/triton/provider.go#L18-L43) uses `SDC_KEY_MATERIAL`, so perhaps we should be compatible with that?

misterbisson commented on triton/access_config.go:
```diff
@@ -75,10 +72,26 @@ func (c *AccessConfig) CreateSDCClient() (*cloudapi.Client, error) {
 	return communicator.Config{}
 }
+
+func processKeyMaterial(keyMaterial string) (string, error) {
```
This would be nice in the Terraform provider as well

ChrisAubuchon committed:
Use SDC_ environment variables in the example

misterbisson commented on README.md:
```diff
@@ -14,25 +14,25 @@ To build from source:
-        "sdc_account": "{{env `SDC_ACCOUNT`}}",
-        "sdc_key_id": "{{env `SDC_KEY_ID`}}",
-        "sdc_key_path": "{{env `SDC_KEY_PATH`}}"
+        "triton_url": "{{env `TRITON_URL`}}",
```
This should match [`example.json`, no](https://github.com/joyent/packer-builder-triton/pull/3/files#diff-964a59d491880661889b0e908c47a1e0R3)?

ChrisAubuchon committed:
Match Readme with example

ChrisAubuchon committed:
Create structure members if they are not set in the configuration
